File: src/main/scala/clase6/ADTUnbundledStatefulOpen.scala
package clase6

object ADTUnbundledStatefulOpen extends App {

  case class Stack[T](var content: List[T])

  def stack[T](): Stack[T] = Stack[T](Nil)

  def push[T](s: Stack[T], element: T) = {
    s.content = element :: s.content
  }

  def pop[T](s: Stack[T]) = {
    s.content match {
      case head :: tail =>
        s.content = tail
        head
    }
  }

  def isEmpty[T](s: Stack[T]) = s.content == Nil

  assert(isEmpty(stack()))

  val s = stack[String]()

  push(s, "ariel")
  push(s, "alejandro")

  assert(!isEmpty(s))

  assert(pop(s) == "alejandro")
  assert(pop(s) == "ariel")
  assert(isEmpty(s))

}

File: haskell/src/clase6/ADTUnbundledStatefulOpen.hs
module ADTUnbundledStatefulOpen where

import Debug.Trace
import Data.IORef

stack = newIORef []

push stack element = modifyIORef stack (element : )

pop stack = do
  current <- readIORef stack
  case current of
    (head : tail) -> do
      writeIORef stack tail
      return head

isEmpty stack = do
  current <- readIORef stack
  return (case current of
    [] -> True
    _ -> False)

main = do
  s <- stack
  isEmpty s >>= traceShowM -- True
  push s "ariel"
  push s "alejandro"
  isEmpty s >>= traceShowM -- False
  pop s >>= traceShowM -- "alejandro"
  pop s >>= traceShowM -- "ariel"
  isEmpty s >>= traceShowM -- True
